packages/lib-classifier/src/store/SubjectStore.spec.js

```diff
@@ -1,147 +1,131 @@
+    store.workflows.setActive(workflow.id)
+    return store
   }
   describe('Actions > advance', function () {
+    let rootStore
     before(function () {
-      rootStore = RootStore.create(
-        { projects: ProjectStore.create(), subjects: SubjectStore.create(), workflows: WorkflowStore.create() },
-        { client: clientStub }
-      )
+      rootStore = setupStores()
     })
 
-    it('should make the next subject in the queue active when calling `advance()`', function (done) {
-      setupStores(rootStore)
-      rootStore.subjects.populateQueue().then(() => {
-        expect(rootStore.subjects.active.id).to.equal(rootStore.subjects.resources.values().next().value.id)
-        rootStore.subjects.advance()
-        expect(rootStore.subjects.active.id).to.equal(rootStore.subjects.resources.values().next().value.id)
-        expect(rootStore.subjects.resources.get('1')).to.be.undefined()
-      }).then(done, done)
+    it('should make the next subject in the queue active when calling `advance()`', function () {
+      expect(rootStore.subjects.active.id).to.equal(rootStore.subjects.resources.values().next().value.id)
+      rootStore.subjects.advance()
+      expect(rootStore.subjects.active.id).to.equal(rootStore.subjects.resources.values().next().value.id)
+      expect(rootStore.subjects.resources.get('1')).to.be.undefined()
     })
 
     describe('with less than three subjects in the queue', function () {
-      let populateSpy
+      let rootStore
 
       before(function () {
-        populateSpy = sinon.spy(rootStore.subjects, 'populateQueue')
+        rootStore = setupStores()
+      })
+
+      beforeEach(function () {
+        rootStore.subjects.populateQueue.resetHistory()
         while (rootStore.subjects.resources.size > 2) {
           rootStore.subjects.advance()
         }
       })
 
+      afterEach(function () {
+      })
+
       after(function () {
         rootStore.subjects.populateQueue.restore()
       })
 
       it('should request more subjects', function () {
-        expect(populateSpy).to.have.been.calledOnce()
+        expect(rootStore.subjects.populateQueue).to.have.been.calledOnce()
+      })
+
+      it('should request more subjects when the initial queue response only has less than 3 subjects', function () {
+        const clientStub = stubPanoptesJs({ workflows: workflow, subjects: shortListSubjects })
+        rootStore = setupStores(clientStub)
+        expect(rootStore.subjects.populateQueue).to.have.been.calledOnce()
       })
     })
 
     describe('after emptying the queue', function () {
+      let rootStore
       before(function () {
-        sinon.stub(clientStub.panoptes, 'get').callsFake(() => Promise.resolve({ body: [] }))
+        const clientStub = stubPanoptesJs({ workflows: workflow, subjects: [] })
+        rootStore = setupStores(clientStub)
+      })
+
+      beforeEach(function () {
         while (rootStore.subjects.resources.size > 0) {
```

Comment:
The queue size is 0 here, because the root store is initialised with an empty queue, so this loop never runs and we don't test what happens when we advance a queue until it is empty.

Can the root store be initialised with subjects in the queue? On master, this test re-uses the root store from thr previous block, which has 2 subjects in the queue.

Reply:
I haven't had much success setting a map via snapshot for test mocks, so it's been easier for me using the actual API of the stores. I've removed the reset history call from the spy to show that the calls happen as expected, once on initialization and once when the queue advances to below three subjects, so the while loop here does work. Instead I've left comments clarifying what's being tested.

Reply:
I'm running this locally, logging the while loop, and it isn't running. The test is setting up an empty queue, skipping the while loop then testing that the empty store is empty. Maybe it should expect `subjects.advance` to have been called at least once?

If I add that expectation, I get this failure.
```sh
1) Model > SubjectStore
       Actions > advance
         after emptying the queue
           should leave the active subject empty:
     AssertionError: expected advance to have been called at least once, but it was never called
      at Context.called (src/store/SubjectStore.spec.js:122:57)

```

Reply:
Sorry, on first read through of your first comment, I missed that the issue was the setup. So I'll update the panoptes js stub to respond with a queue of subjects. Thanks.

Reply:
I think that will break this test because we do want to advance the queue until it’s empty, and make sure that the store doesn’t crash when you advance a queue that has 2 or fewer subjects.

Reply:
Could you explain how it'd break the test? The stores are instantiated with an initial async request to populate the queue, the panoptes js stub is now correctly responding with a queue of subjects and then the while loop is called that calls advance directly, so the queue is being advanced until it is empty before the expectation is run.

Reply:
The while loop will repopulate the queue, when the queue size is less than 3, so that we never reach 0 and don’t test the queue in the case where it isn’t being refilled.

Reply:
If the test was asynchronous, then it would fail because of the reason you've given. I suppose relying on the synchronicity of the current expectation is a bit brittle and unclear.

Reply:
The suggested refactor to separate the populate and advance actions I think will help with this.

Reply:
Yes. What this test should really do is add 2 subjects to the initial queue during setup. 

At the moment it’s setting up the conditions for an infinite loop of `advance` calling `populateQueue` calling `advance` etc.